### ISSUE TITLE
Unauthorized Access Exception

### ISSUE THREAD

**Issue by Snailic** [EmmittJ/PoESkillTree#131]
Error:
System.UnauthorizedAccessException: Access to the path 'C:\Users\JohnB\AppData\Local\assembly\dl3\HGY7ZNX1.84N\W2BNDYZ0.4K3\701161da\99e6e84d_a373d001\POESKillTree.exe.config' is denied.
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.File.InternalCopy(String sourceFileName, String destFileName, Boolean overwrite, Boolean checkHost)
   at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
   at POESKillTree.SkillTreeFiles.Bootstrap.Main(String[] arguments) in c:\Users\SpaceOgre\Documents\GitHub\PoESkillTree\WPFSKillTree\SkillTreeFiles\Bootstrap.cs:line 80

What is happening is that whenever I close POESkillTree, in the __AssemblyInfo__, spaces are being added in between each character of the file path: P O E S K i l l T r e e , 2 . 0 . 0 . 0 , ,   f i l e : / / / F : / P a t h   o f   E x i l e   T o o l s / P o E S k i l l T r e e / P O E S K i l l T r e e . E X E   

and in order to open POESkillTree I have Find&Replace to get rid of all of the empty space.

I am guessing it is being caused by my AntiVirus (BitDefender) as it had popped up believing that POESkillTree appeared to be malicious shortly after I had installed BitDefender.

It still happens even though the paths are listed as exceptions.

To clarify, this issue didn't happen before I installed BitDefender and is only a mild inconvenience but I still want to report the issue to help others who might face the same problem.

Edit: BitDefender not MalwareBytes and still wasn't able to open when BitDefender was turned off. Copying over the config file from the other folder fixed the issue.

**Comment by EmmittJ** [comment]
What version of the planner are you using?

**Comment by Snailic** [comment]
Latest Release: http://i.gyazo.com/52fdb0ecf59853f29257885e0e6524a9.png

**Comment by EmmittJ** [comment]
Ok, so the spaces are normal. The problem is the tree doesn't have permission to copy files, at least that is what it looks like. Try running the program as admin. 

As a side note, I have MaleWarebytes as well, and I do not have this issue.

**Comment by Snailic** [comment]
Tried running it as admin, no bueno, only removing the spaces seems to fix the issue until i close the program again.

**Comment by EmmittJ** [comment]
Hmm, I just threw it on two different HDDs and it seems to be working fine. Try removing the assembly files and seeing what happens. There is a problem where it cannot copy a new file over for some reason.

**Comment by Snailic** [comment]
POESKillTree.exe.config Wasn't copying to 99e6e84d_a373d001 from 99e6e84d_a373d001_0. Copied it over by hand and the issue stopped.

I checked the permissions on the folder and non of them were blocked.

**Comment by l0g0sys** [comment]
AssemblyInfo.ini in assembly shadow cache should contain space after each character as it uses Unicode encoding. But you shouldn't mess with assembly cache at all.
I'm guessing... you have pinned running application and now it launches from cache instead of original location?
Unpin application and pin .exe from original location where it was installed.

Maybe i should disable pinning of running application to avoid this. So people will be forced to pin original .exe manually.

**Comment by Snailic** [comment]
Actually, it isn't pinned at all. For some reason, it just wasn't able to copy over that file.

**Comment by l0g0sys** [comment]
Ah ok, then it's something else.
Can you check please, if any of the files in original PoESkillTree folder is flagged as "This file was downloaded from another computer, blah blah blah" in file properties together with Unblock button?

**Comment by Snailic** [comment]
Yeah, that is the issue. Literally the first time I have ever had an issue like that. Thanks for the help.

**Comment by l0g0sys** [comment]
I gave this issue a bit of thinking in last few days, as this is not first case and won't be last.
As you might know, this issue is caused by downloading released Zip package, and extracting it using Windows native tools (f.e. Explorer). This will cause every extracted file from such downloaded archive to be flagged as "came from another computer", which is done by assigning a security zone identifier to each file using NTFS streams.
```
E:\>more <PoESkillTree-2.1.3.zip:Zone.Identifier
[ZoneTransfer]
ZoneId=3
```
.NET CLR is very sensitive about these zones, and it can block certain assembly functionalities, or even block access to files from different zones depending on policy settings.

So, what can be done? I've got few solutions so far:
  1. Noob way: On each application launch, all important files like .exe.config and DLLs will get scanned and stripped of this zone identifier. Question is, if this stripping will .NET actually allow...
  2. Semi-noob way: Provide self-extracting archive beside Zip archive, which we still need for auto update. This could be done by using 7zip SFX archive mode. And to discourage downloading of Zip archive manually, we could just change it's extension to something like ".pkg".
  3. Pro way: Provide executable with full blown setup wizard, with all the bells and whistles it can provide:
    - Localized setup process with language selection, which could also be used to pre-set application Language option
	- Uninstall method
	- Option to place shortcut on desktop, add to Start menu
	- and all other stuff you are used to when installing some application

	This could be done by either using MS InstallShield, but that would exclude anyone using VS Express edition for development, or by using some free/open source tool such as:
		- WiX Toolset (http://wixtoolset.org/)
		- Inno Setup (http://www.jrsoftware.org/isinfo.php).
		- NSIS - Nullsoft installer (http://nsis.sourceforge.net/Main_Page)

Any other thoughts? :)

**Comment by EmmittJ** [comment]
We should probably have an installer honestly. This project is big enough
that it could use an official setup tool

**Comment by MLanghof** [comment]
I really like portable apps that I can just unpack and go. It feels like I'm adding clutter from installing many tiny applications when they wouldn't actually need much besides the .exe and a nearby folder to save stuff in.

However, that is purely personal taste and probably bears no weight against actual reasons to use an installer.

**Comment by l0g0sys** [comment]
Yep, application data separation is mandatory once you want to install app into system folders like "Program Files".

So, i did some testing and reading, and now i can present installer preview.
Download from my repo:
https://github.com/l0g0sys/PoESkillTree/releases/download/2.1.3/PoESkillTree-2.1.3.exe

The application data are stored in Environment.SpecialFolder.ApplicationData (C:\Users\login\AppData\PoESkillTree), while application itself can be safely installed into Program files.

Still need to solve few issues like:
  - Multiple installations of different versions.
    This means that, when we release PoE 2.0 compatible version, when installed it must be able to coexist with old installed version. The old installed version and new one must use different application data folder, different AppId (so you will see 2 versions installed in Uninstall program control panel), different Start menu group, etc.
  - Localization of setup wizard.
    Inno Setup has own localization files, but i'd like to use centralized localization we use for application, if possible.
  - Automatic updates.
    It should be possible to just download installer with "Check for updates" function and execute it after application exit. This could also make all that shadow copy mess unnecessary.

**Comment by l0g0sys** [comment]
Fixed by #141
